_posts/Article/Engineering/Certbot-issues-wildcards-private-network.md

```diff
@@ -0,0 +1,218 @@
+tags:
+  - https
+  - privatenetwork
+  - devops
+toc: true
+---
+# 使用 Certbot 在内网签发 wildcards（通配符证书）
+
+### 前置条件
+
+1.  Certbot 和 certbot-dns-cloudflare 插件
+2.  域名的 NS 服务器设置为 cloudflare， 配置好，大概一天后生效（保守估计）
+3.  内网 DNS 服务器， 如果路由器支持定义局域网域名，可以不安装，我用的是 adguard
+4.  内网 Ubuntu 服务器  
+5.  Ubuntu 安装 Nginx
+
+#### 安装 Certbot
+
+``` bash
+sudo apt install -y certbot
+```
+
+因为域名的 NS 服务器在 cloudflare， 需要使用 `certbot-dns-cloudflare` 插件
+
+``` shell
+sudo apt-get install python3-certbot-dns-cloudflare
+```
+
+#### 获取用户 API 令牌
+
+- 登录到 Cloudflare Dashboard 。
+- 前往 "My Profile" \> ["API Tokens"](https://dash.cloudflare.com/profile/api-tokens) 。
+- 确认你的 API 令牌具有所需的权限（`Zone.Zone:Read` 和 `Zone.DNS:Edit`），并且选择要使用的域名。
+
+在服务器上创建文件 `vim ~/.cloudflare.ini`,写入如下内容
+
+``` ini
+dns_cloudflare_api_token =  xxx  ##  对应域名的 api token
+```
+
+然后修改文件权限，只可以本用户可读写 `chmod  600 ~/.cloudflare.ini`
+
+#### 签发证书
+
+``` shell
+ sudo certbot certonly \                                                       --dns-cloudflare \                                                       --dns-cloudflare-credentials    ~/.cloudflare.ini  -d  \                 'yourdomain.com, *.yourdomain.com'
+```
+
+就可以看到如下
+
+``` shell
+Saving debug log to /var/log/letsencrypt/letsencrypt.log
+Requesting a certificate for yourdomain.com and *.yourdomain.com
+Waiting 10 seconds for DNS changes to propagate
+
+Successfully received certificate.
+Certificate is saved at: /etc/letsencrypt/live/yourdomain.com/fullchain.pem
+Key is saved at:         /etc/letsencrypt/live/yourdomain.com/privkey.pem
+This certificate expires on 2024-08-10.
+
+These files will be updated when the certificate renews.
+
+Certbot has set up a scheduled task to automatically renew this certificate in the background.
+- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+If you like Certbot, please consider supporting our work by:
+
+ * Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
+
+ * Donating to EFF:                    https://eff.org/donate-le
+
+- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+```
+
+几分钟后应该收到 cloudflare 和 Certbot 的邮件。
+
+<figure>
+<img
+src="Certbot-issues-wildcards-private-network/6af9d79b7229c6957b734c0920e4c1753af89b3a.png"
+title="wikilink" alt="../posts/assets/Pastedimage20240513082444.png" />
+<figcaption
+aria-hidden="true"></figcaption>
+</figure>
+
+因为这个证书是有 90 天的有效期， 需要把 cerbot 加入定时任务。先验证 Cerbot 续签是否可以成功
+
+``` shell
+certbot renew --dry-run
+```
+
+ Ubuntu 和 Debian 使用了 `systemd` ， 在安装 Certbot 过程中已经设置好了，不需要在自己手动加入 `Crontab` ， 验证状态。
+ \`\`\`shell
+ systemctl status certbot.timer
+
+● certbot.timer - Run certbot twice daily Loaded: loaded (/lib/systemd/system/certbot.timer; enabled; preset: enabled) Active: active (waiting) since Sun 2024-05-12 22:49:11 CST; 53min ago Trigger: Mon 2024-05-13 01:49:57 CST; 2h 7min left Triggers: ● certbot.service May 12 22:49:11 traefik systemd\[1\]: Started certbot.timer - Run certbot twice daily.

```

Comment:
```suggestion
● certbot.timer - Run certbot twice daily Loaded: loaded (/lib/systemd/system/certbot.timer; enabled; preset: enabled) Active: active (waiting) since Sun 2024-05-12 22:49:11 CST; 53min ago Trigger: Mon 2024-05-13 01:49:57 CST; 2h 7min left Triggers: ● certbot.service May 12 22:49:11 traefik systemd\[1\]: Started certbot.timer - Run certbot twice daily.
```
这个输出原来是什么样子的？是不是有一些换行？能稍微整理一下吗？

Reply:
转换工具导致出现的问题。